flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/savepoint/SavepointV2.java

```diff
@@ -168,10 +168,27 @@ public static Savepoint convertToOperatorStateSavepointV2(
 				expandedToLegacyIds = true;
 			}
 
+			if (jobVertex == null) {
+				throw new IllegalStateException(
+					"Could not find task for state with ID " + taskState.getJobVertexID() + ". " +
+					"When migrating a savepoint from a version < 1.3 please make sure that the topology was not " +
+					"changed through removal of a stateful operator or modification of a chain containing a stateful " +
+					"operator.");
+			}
+
 			List<OperatorID> operatorIDs = jobVertex.getOperatorIDs();
 
 			for (int subtaskIndex = 0; subtaskIndex < jobVertex.getParallelism(); subtaskIndex++) {
-				SubtaskState subtaskState = taskState.getState(subtaskIndex);
+				SubtaskState subtaskState;
+				try {
+					subtaskState = taskState.getState(subtaskIndex);
```

Comment:
Sorry for commenting late on this but I have had some major migration issues in the last few days :D 
I think we should explicitly compare parallelism instead of relying on the error:
if (taskState.getStates().size() != jobVertex.getParallelism()) --> error

Otherwise this will not fail on lower parallelism.